File: packages/sdk-python/rateguard/core/realtime.py
"""Realtime session enforcement — the voice substrate.

Voice sessions (OpenAI Realtime, Gemini Live) are one WebSocket that can
burn dollars per minute for hours; request-based rate limiting is
structurally blind to them. This module extracts token usage from
realtime SERVER events and budgets a session continuously. It is the
substrate the Pipecat / LiveKit Agents adapters sit on.

Transport-agnostic by design: RateGuard never touches the socket. The
integrator feeds each inbound server frame (a copy — byte transparency
stays with the caller's loop) to the guard and acts on its decision.

Schema provenance, stated precisely because the two differ:
- Gemini Live: LIVE-VERIFIED 2026-07-10 against the real API
  (models/gemini-2.5-flash-native-audio-latest, free tier).
  usageMetadata arrives with the turn-completing message and is
  PER-TURN (verified with a two-turn session: counts do not
  accumulate), with modality-split detail arrays and thoughtsTokenCount.
- OpenAI Realtime: schema-validated against the documented server events
  (response.done carries response.usage); live verification pending —
  no free tier. These counts are estimates vs the billing meter.

Session semantics that follow: usage events are SUMMED per session for
both providers (per-response for OpenAI, per-turn for Gemini) — the
opposite of SSE usage inside one response (MAX-merge), because realtime
events each describe a disjoint slice of work.

Enforcement stance: the guard DECIDES, the integrator ACTS. On first
breach ``on_exceeded`` fires exactly once and the state is terminal —
close the socket with a proper close frame, degrade to text, or
downgrade the model. Frames are never rewritten.
"""

from __future__ import annotations

import json
import threading
import time as _time
from dataclasses import dataclass, field, replace
from typing import Any, Callable, Literal

from .. import types as _types

RealtimeProviderName = Literal["openai", "gemini"]


@dataclass(frozen=True)
class RealtimeUsage:
    """One usage observation. All fields are token counts; 0 = not reported."""

    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    input_text_tokens: int = 0
    input_audio_tokens: int = 0
    input_cached_tokens: int = 0
    output_text_tokens: int = 0
    output_audio_tokens: int = 0
    thoughts_tokens: int = 0  # Gemini's thoughtsTokenCount

    def add(self, other: "RealtimeUsage") -> "RealtimeUsage":
        return RealtimeUsage(
            input_tokens=self.input_tokens + other.input_tokens,
            output_tokens=self.output_tokens + other.output_tokens,
            total_tokens=self.total_tokens + other.total_tokens,
            input_text_tokens=self.input_text_tokens + other.input_text_tokens,
            input_audio_tokens=self.input_audio_tokens + other.input_audio_tokens,
            input_cached_tokens=self.input_cached_tokens + other.input_cached_tokens,
            output_text_tokens=self.output_text_tokens + other.output_text_tokens,
            output_audio_tokens=self.output_audio_tokens + other.output_audio_tokens,
            thoughts_tokens=self.thoughts_tokens + other.thoughts_tokens,
        )


@dataclass(frozen=True)
class RealtimeEvent:
    """Parsed view of one server frame. usage is None when the event
    carries no usage report (most deltas)."""

    provider: RealtimeProviderName
    type: str
    usage: RealtimeUsage | None = None
    turn_complete: bool = False


def parse_openai_realtime_event(raw: bytes | str) -> RealtimeEvent:
    """One OpenAI Realtime server event. Usage rides "response.done"
    (one response = one model turn)."""
    data = json.loads(raw)
    event_type = str(data.get("type", "unknown"))
    out_usage: RealtimeUsage | None = None
    turn_complete = event_type == "response.done"
    if turn_complete:
        u = (data.get("response") or {}).get("usage")
        if u is not None:
            in_details = u.get("input_token_details") or {}
            out_details = u.get("output_token_details") or {}
            out_usage = RealtimeUsage(
                input_tokens=int(u.get("input_tokens", 0)),
                output_tokens=int(u.get("output_tokens", 0)),
                total_tokens=int(u.get("total_tokens", 0)),
                input_text_tokens=int(in_details.get("text_tokens", 0)),
                input_audio_tokens=int(in_details.get("audio_tokens", 0)),
                input_cached_tokens=int(in_details.get("cached_tokens", 0)),
                output_text_tokens=int(out_details.get("text_tokens", 0)),
                output_audio_tokens=int(out_details.get("audio_tokens", 0)),
            )
    return RealtimeEvent(provider="openai", type=event_type, usage=out_usage, turn_complete=turn_complete)


def parse_gemini_live_event(raw: bytes | str) -> RealtimeEvent:
    """One Gemini Live server message. usageMetadata is per-turn —
    verified against the live API (see module docstring)."""
    data = json.loads(raw)
    if data.get("setupComplete") is not None:
        event_type = "setupComplete"
        turn_complete = False
    elif data.get("serverContent") is not None:
        event_type = "serverContent"
        turn_complete = bool(data["serverContent"].get("turnComplete", False))
    else:
        event_type = "unknown"
        turn_complete = False

    out_usage: RealtimeUsage | None = None
    meta = data.get("usageMetadata")
    if meta is not None:
        input_text = input_audio = output_text = output_audio = 0
        for d in meta.get("promptTokensDetails") or []:
            if d.get("modality") == "TEXT":
                input_text += int(d.get("tokenCount", 0))
            elif d.get("modality") == "AUDIO":
                input_audio += int(d.get("tokenCount", 0))
        for d in meta.get("responseTokensDetails") or []:
            if d.get("modality") == "TEXT":
                output_text += int(d.get("tokenCount", 0))
            elif d.get("modality") == "AUDIO":
                output_audio += int(d.get("tokenCount", 0))
        out_usage = RealtimeUsage(
            input_tokens=int(meta.get("promptTokenCount", 0)),
            output_tokens=int(meta.get("responseTokenCount", 0)),
            total_tokens=int(meta.get("totalTokenCount", 0)),
            input_text_tokens=input_text,
            input_audio_tokens=input_audio,
            output_text_tokens=output_text,
            output_audio_tokens=output_audio,
            thoughts_tokens=int(meta.get("thoughtsTokenCount", 0)),
        )
    return RealtimeEvent(provider="gemini", type=event_type, usage=out_usage, turn_complete=turn_complete)


def parse_realtime_event(provider: RealtimeProviderName, raw: bytes | str) -> RealtimeEvent:
    if provider == "openai":
        return parse_openai_realtime_event(raw)
    if provider == "gemini":
        return parse_gemini_live_event(raw)
    raise ValueError(f"rateguard: unknown realtime provider {provider!r}")


@dataclass(frozen=True)
class RealtimeCostRates:
    """Micro-USD per MILLION tokens per class (e.g. $32/M = 32_000_000).
    Caller-priced: realtime pricing drifts too fast to bake in."""

    input_text_per_m_tokens: int = 0
    input_audio_per_m_tokens: int = 0
    input_cached_per_m_tokens: int = 0
    output_text_per_m_tokens: int = 0
    output_audio_per_m_tokens: int = 0

    def cost_micro_usd(self, u: RealtimeUsage) -> int:
        # Cached input priced at its own rate; the un-cached remainder of
        # text input at the text rate. Absent detail splits stay zero —
        # never guess a split the provider didn't report.
        uncached_text = max(u.input_text_tokens - u.input_cached_tokens, 0)
        total = (
            uncached_text * self.input_text_per_m_tokens
            + u.input_cached_tokens * self.input_cached_per_m_tokens
            + u.input_audio_tokens * self.input_audio_per_m_tokens
            + u.output_text_tokens * self.output_text_per_m_tokens
            + u.output_audio_tokens * self.output_audio_per_m_tokens
        )
        return total // 1_000_000


@dataclass(frozen=True)
class RealtimeSessionLimits:
    """Bounds for one session. 0 means unlimited."""

    max_total_tokens: int = 0
    max_audio_tokens: int = 0  # input+output audio — the expensive class
    max_turns: int = 0
    max_duration_seconds: float = 0.0
    max_estimated_cost_micro_usd: int = 0


@dataclass(frozen=True)
class RealtimeDecision:
    """The guard's verdict. exceeded is terminal once true (via
    observe/tick; peek may additionally report a derived, uncommitted
    duration breach)."""

    exceeded: bool
    reason: str  # "total_tokens" | "audio_tokens" | "turns" | "duration" | "cost" | ""
    totals: RealtimeUsage
    turns: int
    estimated_cost_micro_usd: int
    elapsed_seconds: float


class _SystemClock:
    def now(self) -> float:
        return _time.time()


@dataclass
class RealtimeSessionGuardOptions:
    limits: RealtimeSessionLimits = field(default_factory=RealtimeSessionLimits)
    cost_rates: RealtimeCostRates = field(default_factory=RealtimeCostRates)
    # Fires exactly once, on the observation that first breaches a limit.
    # Runs synchronously on the observing thread — keep it short.
    on_exceeded: Callable[[RealtimeDecision], None] | None = None
    clock: "_types.Clock | None" = None


class RealtimeSessionGuard:
    """Accumulates realtime usage for ONE session and enforces its
    limits. Thread-safe; create one per session."""

    def __init__(
        self,
        provider: RealtimeProviderName,
        options: RealtimeSessionGuardOptions | None = None,
    ) -> None:
        self._provider: RealtimeProviderName = provider
        self._opts = options or RealtimeSessionGuardOptions()
        self._clock = self._opts.clock or _SystemClock()
        self._started = self._clock.now()
        self._lock = threading.Lock()
        self._totals = RealtimeUsage()
        self._turns = 0
        self._cost = 0
        self._exceeded = False
        self._reason = ""
        self._notified = False

    def observe_raw(self, raw: bytes | str) -> tuple[RealtimeEvent, RealtimeDecision]:
        """Parse one inbound server frame and feed it to the guard. Raises
        on unparseable frames without corrupting state."""
        ev = parse_realtime_event(self._provider, raw)
        return ev, self.observe_event(ev)

    def observe_event(self, ev: RealtimeEvent) -> RealtimeDecision:
        fire: Callable[[RealtimeDecision], None] | None = None
        with self._lock:
            if ev.usage is not None:
                self._totals = self._totals.add(ev.usage)
                self._cost += self._opts.cost_rates.cost_micro_usd(ev.usage)
            if ev.turn_complete:
                self._turns += 1
            decision, fire = self._commit_locked()
        if fire is not None:
            fire(decision)
        return decision

    def tick(self) -> RealtimeDecision:
        """An observation of nothing but time — for a timer loop enforcing
        max_duration on a quiet session. Mutating like observe_event."""
        with self._lock:
            decision, fire = self._commit_locked()
        if fire is not None:
            fire(decision)
        return decision

    def peek(self) -> RealtimeDecision:
        """Pre-flight verdict: no state change, never fires on_exceeded.
        A not-yet-committed duration breach is REPORTED (derived from the
        clock) but not stored — the next observe/tick commits it."""
        with self._lock:
            elapsed = self._clock.now() - self._started
            decision = self._decision_locked(elapsed)
            if not self._exceeded:
                reason = self._breach_locked(elapsed)
                if reason:
                    decision = replace(decision, exceeded=True, reason=reason)
            return decision

    def _breach_locked(self, elapsed: float) -> str:
        limits = self._opts.limits
        if 0 < limits.max_total_tokens < self._totals.total_tokens:
            return "total_tokens"
        if 0 < limits.max_audio_tokens < self._totals.input_audio_tokens + self._totals.output_audio_tokens:
            return "audio_tokens"
        if 0 < limits.max_turns < self._turns:
            return "turns"
        if 0 < limits.max_duration_seconds < elapsed:
            return "duration"
        if 0 < limits.max_estimated_cost_micro_usd < self._cost:
            return "cost"
        return ""

    def _commit_locked(self) -> tuple[RealtimeDecision, Callable[[RealtimeDecision], None] | None]:
        elapsed = self._clock.now() - self._started
        if not self._exceeded:
            reason = self._breach_locked(elapsed)
            if reason:
                self._exceeded = True
                self._reason = reason
        decision = self._decision_locked(elapsed)
        if self._exceeded and not self._notified and self._opts.on_exceeded is not None:
            self._notified = True
            return decision, self._opts.on_exceeded
        return decision, None

    def _decision_locked(self, elapsed: float) -> RealtimeDecision:
        return RealtimeDecision(
            exceeded=self._exceeded,
            reason=self._reason,
            totals=self._totals,
            turns=self._turns,
            estimated_cost_micro_usd=self._cost,
            elapsed_seconds=elapsed,
        )

File: packages/sdk-node/src/core/realtime.ts
/**
 * Realtime session enforcement — the voice substrate.
 *
 * Voice sessions (OpenAI Realtime, Gemini Live) are one WebSocket that
 * can burn dollars per minute for hours; request-based rate limiting is
 * structurally blind to them. This module extracts token usage from
 * realtime SERVER events and budgets a session continuously.
 *
 * Transport-agnostic: RateGuard never touches the socket. The integrator
 * feeds each inbound server frame to the guard and acts on its decision.
 *
 * Schema provenance:
 * - Gemini Live: LIVE-VERIFIED 2026-07-10 against the real API
 *   (gemini-2.5-flash-native-audio-latest, free tier). usageMetadata is
 *   PER-TURN (proven with a two-turn session), with modality-split
 *   detail arrays and thoughtsTokenCount.
 * - OpenAI Realtime: documented response.done schema; live verification
 *   pending — no free tier. Counts are estimates vs the billing meter.
 *
 * Session semantics: usage events are SUMMED per session for both
 * providers — the opposite of SSE usage inside one response (MAX-merge),
 * because realtime events each describe a disjoint slice of work.
 *
 * Enforcement stance: the guard DECIDES, the integrator ACTS. Terminal
 * on first breach, onExceeded fires exactly once; close the socket with
 * a proper close frame, degrade to text, or downgrade the model. Frames
 * are never rewritten.
 */

import type { Clock } from '../types.js';

export type RealtimeProviderName = 'openai' | 'gemini';

/** One usage observation. All fields are token counts; 0 = not reported. */
export interface RealtimeUsage {
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  inputTextTokens: number;
  inputAudioTokens: number;
  inputCachedTokens: number;
  outputTextTokens: number;
  outputAudioTokens: number;
  /** Gemini's thoughtsTokenCount. */
  thoughtsTokens: number;
}

export function emptyRealtimeUsage(): RealtimeUsage {
  return {
    inputTokens: 0,
    outputTokens: 0,
    totalTokens: 0,
    inputTextTokens: 0,
    inputAudioTokens: 0,
    inputCachedTokens: 0,
    outputTextTokens: 0,
    outputAudioTokens: 0,
    thoughtsTokens: 0,
  };
}

function addUsage(a: RealtimeUsage, b: RealtimeUsage): RealtimeUsage {
  return {
    inputTokens: a.inputTokens + b.inputTokens,
    outputTokens: a.outputTokens + b.outputTokens,
    totalTokens: a.totalTokens + b.totalTokens,
    inputTextTokens: a.inputTextTokens + b.inputTextTokens,
    inputAudioTokens: a.inputAudioTokens + b.inputAudioTokens,
    inputCachedTokens: a.inputCachedTokens + b.inputCachedTokens,
    outputTextTokens: a.outputTextTokens + b.outputTextTokens,
    outputAudioTokens: a.outputAudioTokens + b.outputAudioTokens,
    thoughtsTokens: a.thoughtsTokens + b.thoughtsTokens,
  };
}

/** Parsed view of one server frame; usage is undefined for most deltas. */
export interface RealtimeEvent {
  provider: RealtimeProviderName;
  type: string;
  usage?: RealtimeUsage;
  turnComplete: boolean;
}

function num(v: unknown): number {
  const n = Number(v ?? 0);
  return Number.isFinite(n) ? n : 0;
}

/** One OpenAI Realtime server event; usage rides "response.done". */
export function parseOpenAIRealtimeEvent(raw: string | Buffer): RealtimeEvent {
  const data = JSON.parse(typeof raw === 'string' ? raw : raw.toString('utf8')) as Record<string, unknown>;
  const type = String(data['type'] ?? 'unknown');
  const turnComplete = type === 'response.done';
  let usage: RealtimeUsage | undefined;
  if (turnComplete) {
    const response = (data['response'] ?? {}) as Record<string, unknown>;
    const u = response['usage'] as Record<string, unknown> | undefined;
    if (u) {
      const inD = (u['input_token_details'] ?? {}) as Record<string, unknown>;
      const outD = (u['output_token_details'] ?? {}) as Record<string, unknown>;
      usage = {
        inputTokens: num(u['input_tokens']),
        outputTokens: num(u['output_tokens']),
        totalTokens: num(u['total_tokens']),
        inputTextTokens: num(inD['text_tokens']),
        inputAudioTokens: num(inD['audio_tokens']),
        inputCachedTokens: num(inD['cached_tokens']),
        outputTextTokens: num(outD['text_tokens']),
        outputAudioTokens: num(outD['audio_tokens']),
        thoughtsTokens: 0,
      };
    }
  }
  return usage === undefined
    ? { provider: 'openai', type, turnComplete }
    : { provider: 'openai', type, turnComplete, usage };
}

interface GeminiModalitySpan {
  modality?: string;
  tokenCount?: number;
}

/**
 * One Gemini Live server message. usageMetadata is per-turn — verified
 * against the live API (see module comment).
 */
export function parseGeminiLiveEvent(raw: string | Buffer): RealtimeEvent {
  const data = JSON.parse(typeof raw === 'string' ? raw : raw.toString('utf8')) as Record<string, unknown>;
  let type = 'unknown';
  let turnComplete = false;
  if (data['setupComplete'] !== undefined) {
    type = 'setupComplete';
  } else if (data['serverContent'] !== undefined) {
    type = 'serverContent';
    turnComplete = Boolean((data['serverContent'] as Record<string, unknown>)['turnComplete']);
  }

  let usage: RealtimeUsage | undefined;
  const meta = data['usageMetadata'] as Record<string, unknown> | undefined;
  if (meta) {
    usage = emptyRealtimeUsage();
    usage.inputTokens = num(meta['promptTokenCount']);
    usage.outputTokens = num(meta['responseTokenCount']);
    usage.totalTokens = num(meta['totalTokenCount']);
    usage.thoughtsTokens = num(meta['thoughtsTokenCount']);
    for (const d of (meta['promptTokensDetails'] as GeminiModalitySpan[] | undefined) ?? []) {
      if (d.modality === 'TEXT') usage.inputTextTokens += num(d.tokenCount);
      else if (d.modality === 'AUDIO') usage.inputAudioTokens += num(d.tokenCount);
    }
    for (const d of (meta['responseTokensDetails'] as GeminiModalitySpan[] | undefined) ?? []) {
      if (d.modality === 'TEXT') usage.outputTextTokens += num(d.tokenCount);
      else if (d.modality === 'AUDIO') usage.outputAudioTokens += num(d.tokenCount);
    }
  }
  return usage === undefined
    ? { provider: 'gemini', type, turnComplete }
    : { provider: 'gemini', type, turnComplete, usage };
}

export function parseRealtimeEvent(provider: RealtimeProviderName, raw: string | Buffer): RealtimeEvent {
  if (provider === 'openai') return parseOpenAIRealtimeEvent(raw);
  if (provider === 'gemini') return parseGeminiLiveEvent(raw);
  throw new Error(`rateguard: unknown realtime provider ${String(provider)}`);
}

/**
 * Micro-USD per MILLION tokens per class (e.g. $32/M = 32_000_000).
 * Caller-priced: realtime pricing drifts too fast to bake in.
 */
export interface RealtimeCostRates {
  inputTextPerMTokens?: number;
  inputAudioPerMTokens?: number;
  inputCachedPerMTokens?: number;
  outputTextPerMTokens?: number;
  outputAudioPerMTokens?: number;
}

function costMicroUSD(rates: RealtimeCostRates, u: RealtimeUsage): number {
  // Cached input priced at its own rate; the un-cached remainder of text
  // input at the text rate. Absent detail splits stay zero — never guess
  // a split the provider didn't report.
  const uncachedText = Math.max(u.inputTextTokens - u.inputCachedTokens, 0);
  const total =
    uncachedText * (rates.inputTextPerMTokens ?? 0) +
    u.inputCachedTokens * (rates.inputCachedPerMTokens ?? 0) +
    u.inputAudioTokens * (rates.inputAudioPerMTokens ?? 0) +
    u.outputTextTokens * (rates.outputTextPerMTokens ?? 0) +
    u.outputAudioTokens * (rates.outputAudioPerMTokens ?? 0);
  return Math.floor(total / 1_000_000);
}

/** Bounds for one session. 0/undefined means unlimited. */
export interface RealtimeSessionLimits {
  maxTotalTokens?: number;
  /** input+output audio — the expensive class. */
  maxAudioTokens?: number;
  maxTurns?: number;
  maxDurationMs?: number;
  maxEstimatedCostMicroUSD?: number;
}

/** The guard's verdict. exceeded is terminal once committed. */
export interface RealtimeDecision {
  exceeded: boolean;
  /** "total_tokens" | "audio_tokens" | "turns" | "duration" | "cost" | "" */
  reason: string;
  totals: RealtimeUsage;
  turns: number;
  estimatedCostMicroUSD: number;
  elapsedMs: number;
}

export interface RealtimeSessionGuardOptions {
  limits?: RealtimeSessionLimits;
  costRates?: RealtimeCostRates;
  /**
   * Fires exactly once, on the observation that first breaches a limit.
   * Runs synchronously — keep it short (signal your socket loop).
   */
  onExceeded?: (decision: RealtimeDecision) => void;
  clock?: Clock;
}

/**
 * Accumulates realtime usage for ONE session and enforces its limits.
 * Create one per session.
 */
export class RealtimeSessionGuard {
  private readonly limits: RealtimeSessionLimits;
  private readonly costRates: RealtimeCostRates;
  private readonly onExceeded: ((d: RealtimeDecision) => void) | undefined;
  private readonly clock: Clock;
  private readonly startedMs: number;

  private totals = emptyRealtimeUsage();
  private turns = 0;
  private cost = 0;
  private exceeded = false;
  private reason = '';
  private notified = false;

  constructor(
    private readonly provider: RealtimeProviderName,
    options: RealtimeSessionGuardOptions = {},
  ) {
    this.limits = options.limits ?? {};
    this.costRates = options.costRates ?? {};
    this.onExceeded = options.onExceeded;
    this.clock = options.clock ?? { now: () => Date.now() };
    this.startedMs = this.clock.now();
  }

  /** Parse one inbound server frame and feed it to the guard. Throws on
   * unparseable frames without corrupting state. */
  observeRaw(raw: string | Buffer): { event: RealtimeEvent; decision: RealtimeDecision } {
    const event = parseRealtimeEvent(this.provider, raw);
    return { event, decision: this.observeEvent(event) };
  }

  observeEvent(ev: RealtimeEvent): RealtimeDecision {
    if (ev.usage) {
      this.totals = addUsage(this.totals, ev.usage);
      this.cost += costMicroUSD(this.costRates, ev.usage);
    }
    if (ev.turnComplete) {
      this.turns += 1;
    }
    return this.commit();
  }

  /**
   * An observation of nothing but time — for a timer loop enforcing
   * maxDurationMs on a quiet session. Mutating like observeEvent.
   */
  tick(): RealtimeDecision {
    return this.commit();
  }

  /**
   * Pre-flight verdict: no state change, never fires onExceeded. A
   * not-yet-committed duration breach is REPORTED (derived from the
   * clock) but not stored — the next observe/tick commits it.
   */
  peek(): RealtimeDecision {
    const elapsed = this.clock.now() - this.startedMs;
    const decision = this.decision(elapsed);
    if (!this.exceeded) {
      const reason = this.breach(elapsed);
      if (reason) {
        return { ...decision, exceeded: true, reason };
      }
    }
    return decision;
  }

  private breach(elapsedMs: number): string {
    const l = this.limits;
    if (l.maxTotalTokens && this.totals.totalTokens > l.maxTotalTokens) return 'total_tokens';
    if (l.maxAudioTokens && this.totals.inputAudioTokens + this.totals.outputAudioTokens > l.maxAudioTokens) {
      return 'audio_tokens';
    }
    if (l.maxTurns && this.turns > l.maxTurns) return 'turns';
    if (l.maxDurationMs && elapsedMs > l.maxDurationMs) return 'duration';
    if (l.maxEstimatedCostMicroUSD && this.cost > l.maxEstimatedCostMicroUSD) return 'cost';
    return '';
  }

  private commit(): RealtimeDecision {
    const elapsed = this.clock.now() - this.startedMs;
    if (!this.exceeded) {
      const reason = this.breach(elapsed);
      if (reason) {
        this.exceeded = true;
        this.reason = reason;
      }
    }
    const decision = this.decision(elapsed);
    if (this.exceeded && !this.notified && this.onExceeded) {
      this.notified = true;
      this.onExceeded(decision);
    }
    return decision;
  }

  private decision(elapsedMs: number): RealtimeDecision {
    return {
      exceeded: this.exceeded,
      reason: this.reason,
      totals: { ...this.totals },
      turns: this.turns,
      estimatedCostMicroUSD: this.cost,
      elapsedMs,
    };
  }
}
